### PR TITLE
Intgrtns 199 tld caching improvement

### PR DESCRIPTION
**Pull request by Uditha-05**
Added TLD price download method when "TLD Import & Pricing Sync" button is clicked. 
Added methods to increase memory and execution time to successfully download and import prices.

### CODE REVIEW

File: modules/registrars/openprovider/src/TldPriceCache.php [hunk 1]
```diff
@@ -14,17 +15,43 @@ class TldPriceCache
...
      * @return bool
      */
-    public function has()
+    public function has($params)
     {
-        if(!@is_file($this->getLocation()))
+        $filePath = $this->getLocation();
+        $timePeriodInMinutes = 1440; // 24 hours
+
+        if (file_exists($filePath)) {
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/TldPriceCache.php:23]: I think better to move these logic also to the downloadTldCache function

Comment (Uditha-05) [reply modules/registrars/openprovider/src/TldPriceCache.php]: This change has been added

**Review by sampikaipakan — CHANGES_REQUESTED** [review]

**Comment by sampikaipakan** [comment]
Seems conflicts

File: modules/registrars/openprovider/src/TldPriceCache.php [hunk 2]
```diff
@@ -47,8 +74,23 @@ public function write($content)
...
-            throw new \Exception('Unable to write to ' . $this->getLocation());
+        // if (!file_put_contents($this->getLocation(), $file_content))
+        //     throw new \Exception('Unable to write to ' . $this->getLocation());
+
+        try {
+            $fp = fopen($this->getLocation(),'w');
+            if($fp === false){
+                $errMsg = "ERROR: Error occurred while writing the file. Unable to write to {$this->getLocation()}. Please review file/folder permissions and ensure fopen(), fwrite() functions are allowed";
+                logModuleCall('openprovider', 'file_writing_error', null, $errMsg, null, null);
+                throw new \Exception($errMsg);
```
[the remark this answers is not in the document]
Comment (Uditha-05) [reply modules/registrars/openprovider/src/TldPriceCache.php]: This change has been added

File: modules/registrars/openprovider/src/TldPriceCache.php [hunk 3]
```diff
@@ -14,17 +15,43 @@ class TldPriceCache
...
+        if (empty($this->get()))
             return false;
 
         return true;
     }
 
+    /**
+     * Download the tld_cache.php file.
+     */
+    protected function downloadTldCache($params)
```
[the remark this answers is not in the document]
Comment (Uditha-05) [reply modules/registrars/openprovider/src/TldPriceCache.php]: This change has been added

File: modules/registrars/openprovider/Controllers/System/TldPricingController.php [hunk 4]
```diff
@@ -26,19 +26,63 @@ public function __construct(Core $core)
...
+        $value = (int) $shorthandValue;
+    
+        switch ($lastChar) {
+            case 'g':
+                $value *= 1024;
+            case 'm':
+                $value *= 1024;
+            case 'k':
+                $value *= 1024;
+                break;
```
[the remark this answers is not in the document]
Comment (Uditha-05) [reply modules/registrars/openprovider/Controllers/System/TldPricingController.php:45]: Decided to keep current changes.

File: modules/registrars/openprovider/Controllers/System/TldPricingController.php [hunk 5]
```diff
@@ -26,19 +26,63 @@ public function __construct(Core $core)
...
+            $currentMemoryLimitStr = ini_get('memory_limit');
+            $currentMemoryLimitBytes = $this->convertShorthandToBytes($currentMemoryLimitStr);
+
+            // Compare in bytes
+            $minRequiredBytes = 256 * 1024 * 1024; // 256 MB in bytes
+
+            if ($currentMemoryLimitBytes < $minRequiredBytes) {
+                ini_set('memory_limit', '256M');
+            } 
+            $currentExecutionTime = ini_get('max_execution_time');
```
[the remark this answers is not in the document]
Comment (Uditha-05) [reply modules/registrars/openprovider/Controllers/System/TldPricingController.php]: This change has been added

**Review by sampikaipakan — APPROVED** [review]

**Pull request merged**